### ISSUE TITLE
dart_ping uses the wrong TTL flag on Windows

### ISSUE THREAD

**Issue by edpaulucci** [point-source/dart_ping#9]
In Windows, the TTL syntax in the PING command is -i (lowercase) and non -I (uppercase).
windows_ping.dart - line 36

**Comment by point-source** [comment]
Will fix tomorrow. Thanks for the report!

**Comment by point-source** [comment]
Hmm it seems the code specifies a capital "I" on this line: https://github.com/point-source/dart_ping/blob/main/lib/src/ping/windows_ping.dart#L36

Which version of dart_ping are you using and can you confirm this is the actual issue? You can print the ping command to the console with `Ping({opts here}).command`. This will show the command that will be issued to the host OS with the options you selected.

I will try and get on a windows machine to test this soon.

**Comment by edpaulucci** [comment]
> 
> 
> Hmm it seems the code specifies a capital "I" on this line: https://github.com/point-source/dart_ping/blob/main/lib/src/ping/windows_ping.dart#L36
> 
> Which version of dart_ping are you using and can you confirm this is the actual issue? You can print the ping command to the console with `Ping({opts here}).command`. This will show the command that will be issued to the host OS with the options you selected.
> 
> I will try and get on a windows machine to test this soon.

**Comment by edpaulucci** [comment]
Thank you for the feedback, but there was another problem. I believe it is due to the language of Windows (Brazilian Portuguese). I did some translations but it has an error at the end. In the final statistics.

**Comment by point-source** [comment]
Ah yep, that makes sense. Let me see if I can abstract the regex strings in a new release to make it easier to override. 

Which parts of the final stats aren't getting parsed? The English version doesn't parse all possible outputs right now anyway so I'm not sure if you're missing something cause your OS lang is Portuguese or because it's just not implemented yet.

**Comment by edpaulucci** [comment]
Error displayed at the end: transmitted: 0, received: 0, time: 0, errors: Ping process exited with code -1.
Just one more detail, to work I added the parameter "-4" (force ipv4) in the ping command.

**Comment by point-source** [comment]
Does your system default to ipv6?

Can you post an example of the native ping output from your system?

**Comment by edpaulucci** [comment]
Disparando 192.168.0.1 com 32 bytes de dados:
Resposta de 192.168.0.1: bytes=32 tempo=1ms TTL=64
Resposta de 192.168.0.1: bytes=32 tempo=1ms TTL=64
Resposta de 192.168.0.1: bytes=32 tempo=2ms TTL=64
Resposta de 192.168.0.1: bytes=32 tempo=3ms TTL=64

Estatísticas do Ping para 192.168.0.1:
    Pacotes: Enviados = 4, Recebidos = 4, Perdidos = 0 (0% de
             perda),
Aproximar um número redondo de vezes em milissegundos:
    Mínimo = 1ms, Máximo = 3ms, Média = 1ms

**Comment by edpaulucci** [comment]
ping google.com.br

Disparando google.com.br [2800:3f0:4001:81e::2003] com 32 bytes de dados:
Resposta de 2800:3f0:4001:81e::2003: tempo=16ms
Resposta de 2800:3f0:4001:81e::2003: tempo=22ms
Resposta de 2800:3f0:4001:81e::2003: tempo=14ms
Resposta de 2800:3f0:4001:81e::2003: tempo=14ms

Estatísticas do Ping para 2800:3f0:4001:81e::2003:
    Pacotes: Enviados = 4, Recebidos = 4, Perdidos = 0 (0% de
             perda),
Aproximar um número redondo de vezes em milissegundos:
    Mínimo = 14ms, Máximo = 22ms, Média = 16ms

**Comment by point-source** [comment]
I just pushed version 5.3.0 which forces IPv4 on Windows with the -4 flag. It also now accepts a custom PingParser object to override the string detection and parsing strings so you can add support for almost any language. Let me know if you need help using the new PingParser and I'll see what I can do

**Comment by edpaulucci** [comment]
CODE: 
 Future<void> testPing() async {
    final ping = Ping('google.com', count: 3);
    print('Running command: ${ping.command}');
    ping.stream.listen((event) {
      print(event);
    });
    await Future.delayed(Duration(seconds: 5));
    await ping.stop();
  }

DEBUG:
Running command: ping -w 2.0 -I 255 -4 -n 5 google.com

ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: FormatException: Unexpected extension byte (at offset 2)
#0      _Utf8Decoder.convertChunked (dart:convert-patch/convert_patch.dart:1887:7)
#1      _Utf8ConversionSink.addSlice (dart:convert/string_conversion.dart:314:28)
#2      _Utf8ConversionSink.add (dart:convert/string_conversion.dart:310:5)
#3      _ConverterStreamEventSink.add (dart:convert/chunked_conversion.dart:72:18)
#4      _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:111:24)
#5      _rootRunUnary (dart:async/zone.dart:1362:47)
#6      _CustomZone.runUnary (dart:async/zone.dart:1265:19)
#7      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1170:7)
#8      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
#9      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#10     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:733:19)
#11     _StreamController._add (dart:async/stream_controller.dart:607:7)
#12     _StreamController.add (dart:async/stream_controller.dart:554:5)
#13     _rootRunUnary (dart:async/zone.dart:1362:47)
#14     _CustomZone.runUnary (dart:async/zone.dart:1265:19)
#15     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1170:7)
#16     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
#17     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#18     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:733:19)
#19     _StreamController._add (dart:async/stream_controller.dart:607:7)
#20     _StreamController.add (dart:async/stream_controller.dart:554:5)
#21     _Socket._onData (dart:io-patch/socket_patch.dart:2144:41)
#22     _rootRunUnary (dart:async/zone.dart:1370:13)
#23     _CustomZone.runUnary (dart:async/zone.dart:1265:19)
#24     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1170:7)
#25     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
#26     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#27     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:733:19)
#28     _StreamController._add (dart:async/stream_controller.dart:607:7)
#29     _StreamController.add (dart:async/stream_controller.dart:554:5)
#30     new _RawSocket.<anonymous closure> (dart:io-patch/socket_patch.dart:1680:33)
#31     _NativeSocket.issueReadEvent.issue (dart:io-patch/socket_patch.dart:1192:14)
#32     _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
#33     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)

VERSION:
Flutter 2.0.4 • channel stable • https://github.com/flutter/flutter.git
Framework • revision b1395592de (2 weeks ago) • 2021-04-01 14:25:01 -0700
Engine • revision 2dce47073a
Tools • Dart 2.12.2
dart_ping: ^5.3.0

**Comment by point-source** [comment]
Okay it looks like the TTL flag was actually incorrect. It should be lower case instead of upper case even though the [official Microsoft docs](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/ping) say uppercase. :roll_eyes: 

I just published 5.3.1 which fixes the flag. That said, your stack trace points to a possibly different error. If you don't mind trying 5.3.1 and confirming before I dig further, it would be appreciated. Thanks!

And just to confirm, this test was run on an English language OS or Portugese? If it's Portugese, you'll need to pass a custom parser object into the parser arg when you instantiate Ping().

**Comment by edpaulucci** [comment]
Hello, I tested it again but it still didn't work. Follow my code and the error. Thanks

flutter: PingResponse(seq:32, ip:172.217.28.78, ttl:21, time:113.0 ms)
flutter: PingResponse(seq:32, ip:172.217.28.78, ttl:14, time:113.0 ms)
flutter: PingResponse(seq:32, ip:172.217.28.78, ttl:13, time:113.0 ms)
flutter: PingResponse(seq:32, ip:172.217.28.78, ttl:10, time:113.0 ms)
[ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: FormatException: Unexpected extension byte (at offset 36)
#0      _Utf8Decoder.convertChunked (dart:convert-patch/convert_patch.dart:1887:7)
#1      _Utf8ConversionSink.addSlice (dart:convert/string_conversion.dart:314:28)
#2      _Utf8ConversionSink.add (dart:convert/string_conversion.dart:310:5)
#3      _ConverterStreamEventSink.add (dart:convert/chunked_conversion.dart:72:18)
#4      _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:111:24)
#5      _rootRunUnary (dart:async/zone.dart:1362:47)
#6      _CustomZone.runUnary (dart:async/zone.dart:1265:19)
#7      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1170:7)
#8      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
#9      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#10     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:733:19)
#11     _StreamController._add (dart:async/stream_controller.dart:607:7)
#12     _StreamController.add (dart:async/stream_controller.dart:554:5)
#13     _rootRunUnary (dart:async/zone.dart:1362:47)
#14     _CustomZone.runUnary (dart:async/zone.dart:1265:19)
#15     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1170:7)
#16     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
#17     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#18     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:733:19)
#19     _StreamController._add (dart:async/stream_controller.dart:607:7)
#20     _StreamController.add (dart:async/stream_controller.dart:554:5)
#21     _Socket._onData (dart:io-patch/socket_patch.dart:2144:41)
#22     _rootRunUnary (dart:async/zone.dart:1370:13)
#23     _CustomZone.runUnary (dart:async/zone.dart:1265:19)
#24     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1170:7)
#25     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
#26     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
#27     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:733:19)
#28     _StreamController._add (dart:async/stream_controller.dart:607:7)
#29     _StreamController.add (dart:async/stream_controller.dart:554:5)
#30     new _RawSocket.<anonymous closure> (dart:io-patch/socket_patch.dart:1680:33)
#31     _NativeSocket.issueReadEvent.issue (dart:io-patch/socket_patch.dart:1192:14)
#32     _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
#33     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)

**Comment by edpaulucci** [comment]
My code:
Future<void> testePing() async {
    var responseStr = RegExp(r'Resposta de');
    var responseRgx = RegExp(r'de (.*): bytes=(\d+) tempo=(\d+)ms TTL=(\d+)');
    var summaryStr = RegExp(r'Perdidos:');
    var summaryRgx =
        RegExp(r'Enviados = (\d+), Recebidos = (\d+), Perdidos = (\d+)');
    var timeoutStr = RegExp(r'host unreachable');
    var unknownHostStr =
        RegExp(r'A solicitação ping não pôde encontrar o host');

    final pparser = PingParser(
        responseStr: responseStr,
        responseRgx: responseRgx,
        summaryStr: summaryStr,
        summaryRgx: summaryRgx,
        timeoutStr: timeoutStr,
        unknownHostStr: unknownHostStr);

    final ping = Ping('google.com', count: 5, parser: pparser);
    print('Running command: ${ping.command}');
    ping.stream.listen((event) {
      print(event);
    });
    await Future.delayed(Duration(seconds: 5));
    await ping.stop();
  }

**Comment by edpaulucci** [comment]
One difference I found is that in English, in the summary line, it is:
**Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),**
In Portuguese, the result is not in a single line:
**Pacotes: Enviados = 4, Recebidos = 4, Perdidos = 0 (0% de
             perda),**

Can this influence the result?

**Comment by point-source** [comment]
Ah, it appears that one or more of the characters are not utf-8 compatible. I have published dart_ping 5.4.2 which allows overriding of the character encoding like so:

`final ping = Ping('google.com', count: 5, encoding: Utf8Codec(allowMalformed: true));`
This will ignore characters which it cannot decode.

OR

`final ping = Ping('google.com', count: 5, encoding: Latin1Codec());`
This is an example of an alternative decoder.

Using your example, I managed to get it working like this:

```
Future<void> testePing() async {
    var responseStr = RegExp(r'Resposta de');
    var responseRgx = RegExp(r'de (.*): bytes=(\d+) tempo=(\d+)ms TTL=(\d+)');
    var summaryStr = RegExp(r'Perdidos');
    var summaryRgx =
        RegExp(r'Enviados = (\d+), Recebidos = (\d+), Perdidos = (\d+)');
    var timeoutStr = RegExp(r'host unreachable');
    var unknownHostStr =
        RegExp(r'A solicitação ping não pôde encontrar o host');

    final pparser = PingParser(
        responseStr: responseStr,
        responseRgx: responseRgx,
        summaryStr: summaryStr,
        summaryRgx: summaryRgx,
        timeoutStr: timeoutStr,
        unknownHostStr: unknownHostStr);

    final ping = Ping('google.com', count: 5, parser: pparser, encoding: Utf8Codec(allowMalformed: true));
    print('Running command: ${ping.command}');
    ping.stream.listen((event) {
      print(event);
    });
  }
```
I had to remove the colon (:) from the summaryStr since on my system, it prints an equals sign (=) but yours may not need that. Lmk how it goes.

**Comment by point-source** [comment]
Closing due to lack of activity
